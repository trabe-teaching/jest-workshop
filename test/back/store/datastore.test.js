import createDataStore from "back/store/datastore";

const expectStoreApi = store =>
  expect(
    Object.keys(store)
      .sort()
      .join(" "),
  ).toEqual("all find findAllBy findBy reset update");

describe("back/store/datastore", () => {
  describe("#createDataStore", () => {
    it("creates an store with seeds", () => {
      const store = createDataStore();
      expectStoreApi(store);
      expect(store.all()).toMatchSnapshot();
    });

    describe("store API", () => {
      const store = createDataStore([
        {
          id: 1,
          name: "Luke Skywalker",
          type: "light",
          rank: "master",
          status: "alive",
          email: "luke@ach.to",
        },
        {
          id: 2,
          name: "Yoda",
          type: "light",
          rank: "master",
          status: "ghost",
          email: "yoda@dagobah.com",
        },
      ]);

      it("creates  store", () => {
        expectStoreApi(store);
      });

      describe("#all", () => {
        it("returns all", () => {
          expect(store.all()).toMatchSnapshot();
        });
      });

      describe("#find", () => {
        it("returns one", () => {
          expect(false && store.find(1)).toMatchSnapshot();
        });
      });

      describe("#findAllBy", () => {
        it("returns all by criteria", () => {
          expect(store.findAllBy(u => u.status === "alive")).toMatchSnapshot();
        });
      });

      describe("#findBy", () => {
        it("returns one by criteria", () => {
          expect(store.findBy(u => u.rank === "master")).toMatchSnapshot();
        });
      });

      describe("#update", () => {
        it("updates the store", () => {
          store.update({ ...store.find(2), status: "unknown" });
          expect(store.all()).toMatchSnapshot();
        });
      });

      describe("#reset", () => {
        it("restores the data", () => {
          store.update({ id: 100, name: "new" });
          expect(store.find(100)).not.toBe(undefined);
          store.reset();
          expect(store.find(100)).toBe(undefined);
          expect(store.all()).toMatchSnapshot();
        });
      });
    });
  });
});
