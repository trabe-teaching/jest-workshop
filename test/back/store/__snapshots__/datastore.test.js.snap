// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`back/store/datastore #createDataStore creates an store with seeds 1`] = `
Array [
  Object {
    "email": "luke@ach.to",
    "id": 1,
    "name": "Luke Skywalker",
    "rank": "master",
    "status": "alive",
    "type": "light",
  },
  Object {
    "email": "yoda@dagobah.com",
    "id": 2,
    "name": "Yoda",
    "rank": "master",
    "status": "ghost",
    "type": "light",
  },
  Object {
    "email": "ben.solo@gmail.com",
    "id": 3,
    "master": 1,
    "name": "Ben Solo",
    "rank": "apprentice",
    "status": "alive",
    "type": "light",
  },
  Object {
    "email": "rey@jaku.mail",
    "id": 4,
    "name": "Rey",
    "rank": "potential",
    "status": "alive",
    "type": "light",
  },
  Object {
    "email": "snoke@altavista.com",
    "id": 5,
    "name": "Supreme Leader Snoke",
    "rank": "master",
    "status": "alive",
    "type": "dark",
  },
]
`;

exports[`back/store/datastore #createDataStore store API #all returns all 1`] = `
Array [
  Object {
    "email": "luke@ach.to",
    "id": 1,
    "name": "Luke Skywalker",
    "rank": "master",
    "status": "alive",
    "type": "light",
  },
  Object {
    "email": "yoda@dagobah.com",
    "id": 2,
    "name": "Yoda",
    "rank": "master",
    "status": "ghost",
    "type": "light",
  },
]
`;

exports[`back/store/datastore #createDataStore store API #find returns one 1`] = `false`;

exports[`back/store/datastore #createDataStore store API #findAllBy returns all by criteria 1`] = `
Array [
  Object {
    "email": "luke@ach.to",
    "id": 1,
    "name": "Luke Skywalker",
    "rank": "master",
    "status": "alive",
    "type": "light",
  },
]
`;

exports[`back/store/datastore #createDataStore store API #findBy returns one by criteria 1`] = `
Object {
  "email": "luke@ach.to",
  "id": 1,
  "name": "Luke Skywalker",
  "rank": "master",
  "status": "alive",
  "type": "light",
}
`;

exports[`back/store/datastore #createDataStore store API #reset restores the data 1`] = `
Array [
  Object {
    "email": "luke@ach.to",
    "id": 1,
    "name": "Luke Skywalker",
    "rank": "master",
    "status": "alive",
    "type": "light",
  },
  Object {
    "email": "yoda@dagobah.com",
    "id": 2,
    "name": "Yoda",
    "rank": "master",
    "status": "ghost",
    "type": "light",
  },
]
`;

exports[`back/store/datastore #createDataStore store API #update updates the store 1`] = `
Array [
  Object {
    "email": "luke@ach.to",
    "id": 1,
    "name": "Luke Skywalker",
    "rank": "master",
    "status": "alive",
    "type": "light",
  },
  Object {
    "email": "yoda@dagobah.com",
    "id": 2,
    "name": "Yoda",
    "rank": "master",
    "status": "unknown",
    "type": "light",
  },
]
`;
